### ISSUE TITLE
createFileNodeFromBuffer example within gatsby-source-filesystem README.md is missing name property

### ISSUE THREAD

**Issue by g00glen00b** [gatsbyjs/gatsby#35242]
### Preliminary Checks

- [X] This issue is not a duplicate. Before opening a new issue, please search existing issues: https://github.com/gatsbyjs/gatsby/issues
- [X] This issue is not a question, feature request, RFC, or anything other than a bug report. Please post those things in GitHub Discussions: https://github.com/gatsbyjs/gatsby/discussions

### Summary

Within the [gatsby-source-filesystem README.md](https://github.com/gatsbyjs/gatsby/tree/master/packages/gatsby-source-filesystem/README.md), the following example is documented for `createFileNodeFromBuffer()`:

```javascript
createFileNodeFromBuffer({
  buffer: value,
  getCache: ctx.getCache,
  createNode: ctx.createNode,
  createNodeId: ctx.createNodeId,
})
```

This example no longer works since Gatsby **v4** because the `name` property is required. This is due to changes within the `createFilePath()` function.

Originally, it contained the [following logic](https://github.com/gatsbyjs/gatsby/blob/1a9469d67a19007faebebfb8ce876970c5e0ffaf/packages/gatsby-source-filesystem/src/utils.js#L52-L54):

```javascript
path.join(directory, `${filename}${ext}`)
```

This meant that even if `filename` was `undefined` , it would generate a file with the name "undefined" (not ideal, but it works).

Nowadays, it contains [some additional logic](https://github.com/gatsbyjs/gatsby/blob/82e3c8a9e17e20ad80483ec277f55f175eca6067/packages/gatsby-core-utils/src/filename-utils.ts#L50-L66). The most important change being:

```javascript
const purgedFileName = filename.replace(filenamePurgeRegex, `-`)
```

This causes a failure if the `name` property is missing within the `createFileNodeFromBuffer()` because it would call `undefined.replace()`.

### Steps to Resolve this Issue

1. Update the `createFileNodeFromBuffer()` example within the [README.md](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/README.md) to include a `name` property, eg.:

```javascript
createFileNodeFromBuffer({
  buffer: value,
  getCache: ctx.getCache,
  createNode: ctx.createNode,
  createNodeId: ctx.createNodeId,
  name: 'filename',
})
```

**Comment by tyhopp** [comment]
Hi @g00glen00b, thanks for the excellent issue description.

Looking at the function interface of `createFileNodeFromBuffer`, the `name` argument has a [default value of `hash`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/src/create-file-node-from-buffer.js#L128) and both `name` and `hash` are [optional in the TS interface definition](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/index.d.ts#L47-L58).

Since we do handle the case of [no `hash`](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/src/create-file-node-from-buffer.js#L155-L157), I'd propose that we add another conditional afterwards to handle no `name`. Something like:

```javascript
if (!hash) {
  hash = createContentDigest(buffer)
}

if (!name) {
  name = hash
}
```

This would allow us to avoid changing the public interface and solve the (probably) unintended behaviour of having files named `undefined` like you mentioned in your description. The fact that the `name` value has a default value of `hash` makes me suspect that this was the original intention as well.

If we did do this then we would need to note this in the README and add a [test case for it](https://github.com/gatsbyjs/gatsby/blob/master/packages/gatsby-source-filesystem/src/__tests__/create-file-node-from-buffer.js) as well.

Would you be interested in taking this up with a PR? I'd be happy to help review.

**Comment by g00glen00b** [comment]
@tyhopp I'll try to make a PR later today.

**Comment by g00glen00b** [comment]
@tyhopp  I created a PR: https://github.com/gatsbyjs/gatsby/pull/35243
In this PR I did the following:

1. I wrote two tests (one testing the behavior if only "hash" was given and one testing the behavior if neither "hash" nor "name" was given).
2. I added the default after the hash is calculated
3. I removed the default within the function arguments as that is no longer necessary
4. I documented the behavior of the `name` and `hash` arguments within the README.md

One thing I wasn't sure about within the tests is how to properly write an assertion based on the generated hash. I ended up importing `createContentDigest()` and calling that within my test to assert that the filename matches the generated hash. I thought about mocking the `createContentDigest()` function, but that would probably cause problems with the other tests.

```javascript
const buffer = createMockBuffer(`buffer-content`)
const expectedHash = createContentDigest(buffer)
await setup({
  buffer: buffer,
  getCache: () => createMockCache(),
})

expect(outputFilename).toContain(`${expectedHash}.bin`)
```
